From 4adbe8f6fa1911fb034e26bec380afd41d10262d Mon Sep 17 00:00:00 2001
From: Dhanavandhana Kannan <quic_dhanavan@quicinc.com>
Date: Tue, 28 Nov 2023 15:13:57 +0530
Subject: [PATCH] wifi: ath12k: Fix adding a QoS control right after dot11
 header

The EAPOL 2/4 packets are dropped in the mac80211 driver
due to incorrect placement of the QoS control in the QoS data frame.

For QoS data, the length of the ieee80211 header includes QoS control
(2 bytes) and HT control (4 bytes). In order to build the
ieee80211 QoS header, the QoS control field should be placed
in the appropriate position in the header.
Currently, when there is an HT control field in the header,
the QoS control field is added to the end which belonging to the
ht control field causing corruption of the header,
which causing EAPOL 2/4 packets to be dropped in mac80211.

To fix, use ieee80211_get_qos_ctl() to get the QoS control pointer
instead of using hdr_len.

Signed-off-by: Ramasamy Kaliappan <quic_rkaliapp@quicinc.com>
Signed-off-by: Dhanavandhana Kannan <quic_dhanavan@quicinc.com>
---
 dp_rx.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

--- a/drivers/net/wireless/ath/ath12k/dp_rx.c
+++ b/drivers/net/wireless/ath/ath12k/dp_rx.c
@@ -2751,7 +2751,7 @@ static void ath12k_get_dot11_hdr_from_rx
 			qos_ctl |= IEEE80211_QOS_CTL_MESH_CONTROL_PRESENT;
 
 		/* TODO: Add other QoS ctl fields when required */
-		memcpy(msdu->data + (hdr_len - IEEE80211_QOS_CTL_LEN),
+		memcpy(ieee80211_get_qos_ctl(msdu->data),
 		       &qos_ctl, IEEE80211_QOS_CTL_LEN);
 	}
 }
